### ISSUE TITLE
verbose-file-reads false positive

### ISSUE THREAD

**Issue by MikailBag** [rust-lang/rust-clippy#5368]
Reproducer on playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a06d5c34e0ec3cd9eafbd42af458028f

std::fs::read only supports reading from path, not from raw handle.

**Comment by MikailBag** [comment]
Second bug: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8e49e12c99d43fe69dbfdbe375f785bf
If this code will be rewritten using `std::fs::read`, it won't be able to reuse one allocation, because `std::fs::read` allocates and returns new buffer

**Comment by MikailBag** [comment]
Third bug: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=9f18b7ce2a3758ff5d37cd4e08ca4afe
Current behavior: all file are read into same buffer, so their contents are concatenated easily.
With std::fs::read: I need manually append.
It is less efficient and it is not less verbose

**Comment by flip1995** [comment]
I'd move this lint to `nursery`, but since these bugs are hard to fix, and therefore this lint probably won't get out of `nursery` any time soon. But since I think the lint idea is good, I'll move it to `restriction`, so it can be enabled if someone wants to check their code about this.

**Comment by vorner** [comment]
I would like to add another false positive to this lint. Here the provided hint doesn't really make sense at all:

```rust
use std::fs::File;
use std::io::Read;
fn main() {
    let mut f = File::open("example.txt").unwrap();
    let mut header = [0; 4];
    f.read_exact(&mut header).unwrap();
    let mut rest = Vec::new();
    f.read_to_end(&mut rest).unwrap();
}
```

Further similar ones would include seeking around the file and reading it multiple times, opening the file sooner, then deleting it on the FS but keeping it open and similar cases.

**Comment by nazar-pc** [comment]
Another false-positive, this time with clearly non-default opening arguments, so simple `fs::read()` is not equivalent:
```rust
#![warn(clippy::verbose_file_reads)]

use std::fs::OpenOptions;
use std::io::Read;

fn main() {
    let mut file = OpenOptions::new()
        .read(true)
        .write(true)
        .create(true)
        .truncate(false)
        .open("/foo")
        .unwrap();
    let mut buf = Vec::new();
    
    file.read_to_end(&mut buf).unwrap();
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=bb6e2b81e9e691033f9f419d3135e696
